### PR TITLE
Refactoring value classes

### LINKED ISSUE #595: Value classes to represent Symbol and SymbolProblem
Towards #574
go/linkage-check-database

### PR DESCRIPTION
**Pull request by suztomo**
Fixes #595 

As per discussion with Elliotte, this change does not use AutoValue, because it causes problems in his Eclipse (#508).

Starting with minimal code: no toString, no builder, no null-check.

### CODE REVIEW

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassInJarTest.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  @Test
+  public void testCreation() {
+    ClassInJar classInJar = new ClassInJar(Paths.get("foo", "bar.jar"), "com.test.Foo");
+    assertEquals("com.test.Foo", classInJar.getClassName());
+    assertEquals(Paths.get("foo", "bar.jar"), classInJar.getJar());
+  }
+
+  @Test
+  public void testEquality() {
+    new EqualsTester()
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassInJarTest.java]: This class is @Beta. I wonder if we can get Guava to commit to this?

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/FieldSymbolTest.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+package com.google.cloud.tools.opensource.classpath;
+
+import static org.junit.Assert.assertEquals;
+
+import com.google.common.testing.EqualsTester;
+import org.junit.Test;
+
+public class FieldSymbolTest {
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/FieldSymbolTest.java]: Need test for nullability or lack thereof.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassSymbolTest.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+package com.google.cloud.tools.opensource.classpath;
+
+import static org.junit.Assert.assertEquals;
+
+import com.google.common.testing.EqualsTester;
+import org.junit.Test;
+
+public class ClassSymbolTest {
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassSymbolTest.java]: Need test for nullability or lack thereof.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassInJarTest.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+import static org.junit.Assert.assertEquals;
+
+import com.google.common.testing.EqualsTester;
+import java.nio.file.Paths;
+import org.junit.Test;
+
+public class ClassInJarTest {
+  @Test
+  public void testCreation() {
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassInJarTest.java]: Need test for nullability or lack thereof.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/SymbolProblemTest.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+package com.google.cloud.tools.opensource.classpath;
+
+import static org.junit.Assert.assertEquals;
+
+import com.google.cloud.tools.opensource.classpath.SymbolProblem.Reason;
+import com.google.common.testing.EqualsTester;
+import java.nio.file.Paths;
+import org.junit.Test;
+
+public class SymbolProblemTest {
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/SymbolProblemTest.java]: Need test for nullability or lack thereof.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/SymbolProblemTest.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+  @Test
+  public void testCreation() {
+    SymbolProblem symbolProblem =
+        new SymbolProblem(
+            new ClassSymbol("java.lang.Integer"),
+            Reason.CLASS_NOT_FOUND,
+            new ClassInJar(Paths.get("foo", "bar.jar"), "java.lang.Object"));
+    assertEquals(Reason.CLASS_NOT_FOUND, symbolProblem.getReason());
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/SymbolProblemTest.java]: assertSame

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+ * source class of the reference.
+ *
+ * @see <a href="https://github.com/GoogleCloudPlatform/cloud-opensource-java/blob/master/library-best-practices/glossary.md#linkage-error">
+ * Java Dependency Glossary: Linkage Error</a>
+ */
+class SymbolProblem {
+
+  private final Reason reason;
+  private final Symbol symbol;
+  private final ClassInJar targetClass;
```
[the remark this answers is not in the document]
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java]: I'd keep the word "target" with regard to the source and target of a reference. "Expected" means something more. For example: _this class expected class `com.google.Foo` from guava 23.0 but it came from guava 19.0._

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    SymbolProblem that = (SymbolProblem) o;
+    return reason == that.reason
```
[the remark this answers is not in the document]
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java]: reason is not nullable. targetClass may be null.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        && Objects.equals(targetClass, that.targetClass);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(reason, symbol, targetClass);
+  }
+
+  /** The kind of linkage error against a symbol reference. */
+  enum Reason {
```
[the remark this answers is not in the document]
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java]: Moved this to outer type.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassInJarTest.java (hunk 1 above)
Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassInJarTest.java]: I just asked Guava repo. https://github.com/google/guava/issues/3452

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassSymbolTest.java (hunk 3 above)
Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassSymbolTest.java]: Added NullPointerTester

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassInJarTest.java (hunk 4 above)
Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassInJarTest.java]: Added NullPointerTester

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/SymbolProblemTest.java (hunk 6 above)
Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/SymbolProblemTest.java]: Updated.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassInJar.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+package com.google.cloud.tools.opensource.classpath;
+
+import java.nio.file.Path;
+import java.util.Objects;
+
+/**
+ * A class in a jar file to uniquely locate the class implementation in a class path.
+ */
+class ClassInJar {
```
[the remark this answers is not in the document]
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassInJar.java]: Updated

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassSymbol.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.opensource.classpath;
+
+import java.util.Objects;
+
+/** Symbol for a class. */
+class ClassSymbol extends Symbol {
```
[the remark this answers is not in the document]
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassSymbol.java]: Updated

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassSymbol.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+  private final String className;
+
+  ClassSymbol(String className) {
+    this.className = className;
```
[the remark this answers is not in the document]
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassSymbol.java]: No, it's not nullable. I added `checkNotNull`.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+   * reason is {@code CLASS_NOT_FOUND}).
+   *
+   * <p>In case of an inner class is missing while its outer class is found in the class path, this
+   * method returns the outer class.
+   */
+  Optional<ClassInJar> getTargetClass() {
+    if (targetClass == null) {
+      return Optional.empty();
+    }
+    return Optional.of(targetClass);
```
[the remark this answers is not in the document]
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java]: Good to know the method. Thanks.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+
+  /**
+   * Returns an {@code Optional} describing the referenced class of the linkage conflict. An empty
+   * {@code Optional} when the target class is not found in the class path (this is the case if
+   * reason is {@code CLASS_NOT_FOUND}).
+   *
+   * <p>In case of an inner class is missing while its outer class is found in the class path, this
+   * method returns the outer class.
+   */
+  Optional<ClassInJar> getTargetClass() {
```
[the remark this answers is not in the document]
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java]: Updated to `@nullable`

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/FieldSymbolTest.java (hunk 2 above)
Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/FieldSymbolTest.java]: Added NullPointerTester

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/SymbolProblemTest.java (hunk 5 above)
Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/SymbolProblemTest.java]: Added NullPointerTester

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Symbol.java [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+ */
+abstract class Symbol {
+  /**
+   * Returns the binary name of the class that contains the symbol. If this is a class symbol, the
+   * class name itself.
+   *
+   * @see <a href="https://docs.oracle.com/javase/specs/jls/se8/html/jls-13.html#jls-13.1">Java
+   *     Language Specification: 13.1. The Form of a Binary</a>
+   */
+  abstract String getClassName();
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Symbol.java]: This doesn't need to be abstract. You could store the className field in this class and set it through the constructor in subclasses. This would avoid a bit of duplicate code.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassInJar.java [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+package com.google.cloud.tools.opensource.classpath;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.nio.file.Path;
+import java.util.Objects;
+
+/** A class in a jar file to uniquely locate the class implementation in a class path. */
+final class ClassInJar {
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassInJar.java]: I'm finding the name "ClassInJar" confusing. Any other iodeas for this object? ClassLocation?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Symbol.java (hunk 15 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Symbol.java]: Updated.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassInJar.java (hunk 16 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassInJar.java]: Simply ClassAndJar?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/FieldSymbol.java [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.Objects;
+
+/** Symbol for a field of a class. */
+final class FieldSymbol extends Symbol {
+  private final String name;
+  private final String descriptor;
+
+  FieldSymbol(String className, String name, String descriptor) {
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/FieldSymbol.java]: Does it work to store just the single binary name of each symbol, rather than splitting it into className and member name?

Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/FieldSymbol.java]: It would work. However, the caller have to combine them when it creates this instance and the user of the data later splits the field into className and field.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/FieldSymbol.java [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+
+  @Override
+  public boolean equals(Object other) {
+    if (this == other) {
+      return true;
+    }
+    if (other == null || getClass() != other.getClass()) {
+      return false;
+    }
+    if (!super.equals(other)) {
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/FieldSymbol.java]: This seems wrong. equals for superclasses is tricky, but this suggests a field can equal a method.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Symbol.java [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+   *
+   * @see <a href="https://docs.oracle.com/javase/specs/jls/se8/html/jls-13.html#jls-13.1">Java
+   *     Language Specification: 13.1. The Form of a Binary</a>
+   */
+  String getClassName() {
+    return className;
+  }
+
+  @Override
+  public boolean equals(Object other) {
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Symbol.java]: This doesn't seem right.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/FieldSymbol.java (hunk 18 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/FieldSymbol.java]: MethodSymbol and FieldSymbol are distinguished 4 lines above by `getClass`.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Symbol.java (hunk 19 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Symbol.java]: It works fine with `EqualsTester`. For clarity, added `abstract` to Symbol class declaration.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/SymbolProblemTest.java [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+            new ClassSymbol("java.lang.Integer"),
+            Reason.CLASS_NOT_FOUND,
+            new ClassAndJar(Paths.get("foo", "bar.jar"), "java.lang.Object"));
+    assertSame(Reason.CLASS_NOT_FOUND, symbolProblem.getReason());
+    assertEquals(new ClassSymbol("java.lang.Integer"), symbolProblem.getSymbol());
+    assertEquals(
+        new ClassAndJar(Paths.get("foo", "bar.jar"), "java.lang.Object"),
+        symbolProblem.getTargetClass());
+
+    new NullPointerTester()
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/SymbolProblemTest.java]: This should be a separate test method.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/MethodSymbolTest.java [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+            "java.lang.Object",
+            "equals",
+            "(Lcom/google/protobuf/Message;)Lio/grpc/MethodDescriptor$Marshaller;");
+    assertEquals("java.lang.Object", methodSymbol.getClassName());
+    assertEquals("equals", methodSymbol.getName());
+    assertEquals(
+        "(Lcom/google/protobuf/Message;)Lio/grpc/MethodDescriptor$Marshaller;",
+        methodSymbol.getDescriptor());
+
+    new NullPointerTester().testConstructors(MethodSymbol.class, Visibility.PACKAGE);
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/MethodSymbolTest.java]: This should be a separate test method. Common code can be pulled up into a fixture (setUp method + fields)

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/FieldSymbolTest.java [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+
+  @Test
+  public void testFieldSymbolCreation() {
+    FieldSymbol fieldSymbol = new FieldSymbol("java.lang.Integer", "MAX_VALUE", "I");
+    assertEquals("java.lang.Integer", fieldSymbol.getClassName());
+    assertEquals("MAX_VALUE", fieldSymbol.getName());
+    assertEquals("I", fieldSymbol.getDescriptor());
+
+    new NullPointerTester().testConstructors(FieldSymbol.class, Visibility.PACKAGE);
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/FieldSymbolTest.java]: This should be a separate test method. Common code can be pulled up into a fixture (setUp method + fields)

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassSymbolTest.java [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+
+  @Test
+  public void testClassSymbolCreation() {
+    ClassSymbol classSymbol = new ClassSymbol("java.lang.Object");
+    assertEquals("java.lang.Object", classSymbol.getClassName());
+
+    new NullPointerTester().testConstructors(ClassSymbol.class, Visibility.PACKAGE);
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassSymbolTest.java]: This should be a separate test method. Common code can be pulled up into a fixture (setUp method + fields). This method might be testGetClassName

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassInJarTest.java [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+    ClassAndJar classInJar = new ClassAndJar(Paths.get("foo", "bar.jar"), "com.test.Foo");
+    assertEquals("com.test.Foo", classInJar.getClassName());
+    assertEquals(Paths.get("foo", "bar.jar"), classInJar.getJar());
+
+    new NullPointerTester().testConstructors(ClassAndJar.class, Visibility.PACKAGE);
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassInJarTest.java]: This should be a separate test method. Common code can be pulled up into a fixture (setUp method + fields)

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/FieldSymbol.java (hunk 18 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/FieldSymbol.java]: That makes sense. It still feels unusual to split the equals logic between the superclass and the subclass. Consider whether equals and hashCode in the superclass could be abstract with all logic in the subclasses.

Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/FieldSymbol.java]: I'd keep as it is (IDE-generated); I considered manually adding className to equals and hashcode in subclasses, but I found:
- The current code encapsulates equals logic (`className`) well within its owning class.
- The current code is straightforward enough to be auto-generated by IDE.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassInJarTest.java (hunk 24 above)
Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassInJarTest.java]: Done (with no common code).

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassSymbolTest.java (hunk 23 above)
Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassSymbolTest.java]: Done (with no common code).

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/FieldSymbolTest.java (hunk 22 above)
Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/FieldSymbolTest.java]: Done (with no common code).

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/MethodSymbolTest.java (hunk 21 above)
Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/MethodSymbolTest.java]: Done (with no common code).

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/SymbolProblemTest.java (hunk 20 above)
Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/SymbolProblemTest.java]: Done (with no common code).

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassAndJar.java [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.nio.file.Path;
+import java.util.Objects;
+
+/**
+ * A tuple of a class name and a jar file to uniquely locate the class implementation in a class
+ * path.
+ */
+final class ClassAndJar {
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassAndJar.java]: I know I wrote this on the whiteboard, but this name suggests an arbitrary pair. Really, this is a class file somewhere, instead of just the name of a class. What about calling it `ClassFile`? It would be logically the same even if it were a .class file somewhere on the filesystem directly instead of in a JAR. You don't need to generalize it to support that case yet.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassAndJar.java [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+  private final Path jar;
+  private final String className;
+
+  ClassAndJar(Path jar, String className) {
+    this.jar = checkNotNull(jar);
+    this.className = checkNotNull(className);
+  }
+
+  /** Returns jar file containing the class. */
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassAndJar.java]: `Returns the path to the JAR file containing the class.`

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassAndJar.java [hunk 27]
```diff
@@ -0,0 +1,63 @@
...
+    }
+    ClassAndJar that = (ClassAndJar) other;
+    return Objects.equals(jar, that.jar) && Objects.equals(className, that.className);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(jar, className);
+  }
+}
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassAndJar.java]: Why not use AutoValue? Seems like a clear win: Reduce this whole class to

```java
@AutoValue
abstract class ClassAndJar {
  abstract Path getJar();
  abstract String getClassName();

  static ClassAndJar create(Path jar, String className) {
    return new AutoValue_ClassAndJar(jar, className);
  }
}
```

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassAndJar.java [hunk 28]
```diff
@@ -0,0 +1,63 @@
...
+    if (this == other) {
+      return true;
+    }
+    if (other == null || getClass() != other.getClass()) {
+      return false;
+    }
+    ClassAndJar that = (ClassAndJar) other;
+    return Objects.equals(jar, that.jar) && Objects.equals(className, that.className);
+  }
+
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassAndJar.java]: `toString()` will be helpful as well.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/FieldSymbol.java (hunk 18 above)
Comment (netdpb) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/FieldSymbol.java]: AutoValue would help immensely here. It also means that as you change the properties you don't have to remember to keep the implementation of this and `hashCode()` up to date. It's a very common source of bugs.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/MethodSymbol.java [hunk 29]
```diff
@@ -0,0 +1,71 @@
...
+  /**
+   * Returns the descriptor of the method. A descriptor holds type information for its parameters
+   * and return value. Example: '{@code
+   * (Lcom/google/protobuf/Message;)Lio/grpc/MethodDescriptor$Marshaller;}', where {@code Message}
+   * class is the parameter and {@code MethodDescriptor$Marshaller} class is the return type.
+   *
+   * @see <a href="https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.3.3">Java
+   *     Virtual Machine Specification: Method Descriptors</a>
+   */
+  String getDescriptor() {
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/MethodSymbol.java]: Do all `Symbol`s have a descriptor? Maybe a `ClassSymbol`'s name should be its canonical name and its descriptor its binary name?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Reason.java [hunk 30]
```diff
@@ -0,0 +1,49 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.opensource.classpath;
+
+/** The kind of linkage error against a symbol reference. */
+enum Reason {
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Reason.java]: Would ErrorType be a better name here? "Reason" is very general, and doesn't even suggest anything negative.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Reason.java [hunk 31]
```diff
@@ -0,0 +1,49 @@
...
+  /** The target class of the symbol reference is not found in the class path. */
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Reason.java]: How is this related to `SYMBOL_NOT_FOUND`? Aren't classes also symbols?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Reason.java [hunk 32]
```diff
@@ -0,0 +1,49 @@
...
+  CLASS_NOT_FOUND,
+
+  /**
+   * The referenced class or interface found in the class path is not compatible with the source.
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Reason.java]: In what ways can it be incompatible?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Reason.java [hunk 33]
```diff
@@ -0,0 +1,49 @@
...
+   */
+  INCOMPATIBLE_CLASS_CHANGE,
+
+  /**
+   * The target class of the symbol reference is inaccessible to the source.
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Reason.java]: How is this materially different from `INACCESSIBLE_SYMBOL`? It's the same, just applied to `MethodSymbol` or `FieldSymbol`, right? Do we really need a separate enum value?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Symbol.java (hunk 19 above)
Comment (netdpb) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Symbol.java]: If you're doing this it means you're comparing this to other and other to null and the classes twice in many cases.

Again, using AutoValue would obviate all this.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java [hunk 34]
```diff
@@ -0,0 +1,84 @@
...
+
+package com.google.cloud.tools.opensource.classpath;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.Objects;
+import javax.annotation.Nullable;
+
+/**
+ * Linkage error caused by a symbol reference that cannot be resolved. This class does not carry the
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java]: Rather than calling this an error caused by a reference, how about just saying it's "A missing or incompatible symbol."

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java [hunk 35]
```diff
@@ -0,0 +1,84 @@
...
+    return reason;
+  }
+
+  /** Returns the target symbol that was not resolved. */
+  Symbol getSymbol() {
+    return symbol;
+  }
+
+  /**
+   * Returns the referenced class of the linkage conflict. Null when the target class is not found
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java]: Did you consider making this throw `ISE` when it's null?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java [hunk 36]
```diff
@@ -0,0 +1,84 @@
...
+   * in the class path (this is the case if the reason is {@code CLASS_NOT_FOUND}).
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java]: Hmm. Even for a nested class?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java [hunk 37]
```diff
@@ -0,0 +1,84 @@
...
+   *
+   * <p>In case of an inner class is missing while its outer class is found in the class path, this
+   * method returns the outer class.
+   */
+  @Nullable
+  ClassAndJar getTargetClass() {
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java]: I wouldn't use the term "target" here. It's the class that should contain the symbol (if it's missing) or does contain the incompatible symbol.

What about `getContainingClass()`?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java [hunk 38]
```diff
@@ -0,0 +1,84 @@
...
+    }
+    if (other == null || getClass() != other.getClass()) {
+      return false;
+    }
+    SymbolProblem that = (SymbolProblem) other;
+    return reason == that.reason
+        && symbol.equals(that.symbol)
+        && Objects.equals(targetClass, that.targetClass);
+  }
+
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java]: What about toString()?

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassInJarTest.java (hunk 1 above)
Comment (netdpb) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/ClassInJarTest.java]: It's not important, since this is a test of your code and you do not distribute it.

That should be better documented in Guava.

**Comment by suztomo** [comment]
@netdpb Thank you for comment. I will incorporate the comments in upcoming PRs (#599). 

I also strongly agree that AutoValue is appropriate for the classes (null check, builders with explicit arguments, toString, equality, and test codes). However, as per Elliotte, AutoValue does not work well in Eclipse and causes him to spend time to fix; that's why this PR is not using AutoValue.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Reason.java (hunk 30 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Reason.java]: Updating to ErrorType: https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/598/files#diff-439f65974a3849e4bec134b2b651f64fR20

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Reason.java (hunk 31 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Reason.java]: It used to be an attribute of a (class, field or method) reference. For example:
- CLASS_NOT_FOUND for a method reference means the target class is missing.
- SYMBOL_NOT_FOUND for a method reference means the method matching name and descriptor is missing.

The reporting may change during the course of this refactoring, but for now I'd keep them as they are.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Reason.java (hunk 32 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Reason.java]: Will add description.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Reason.java (hunk 33 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/Reason.java]: Example difference for a method reference:
- INACCESSIBLE_CLASS for a method reference means the target class is not accessible.
- INACCESSIBLE_SYMBOL for a method reference means the method matching name and descriptor is not accessible.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java (hunk 38 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java]: Adding toString: https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/598/files#diff-f81d38e11d3d9aa7ef6149c9be9aafbaR88

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassAndJar.java (hunk 26 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassAndJar.java]: Updating that: https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/598/files#diff-dcec2e6026c85f37b7538c5c3313b146R38

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassAndJar.java (hunk 27 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassAndJar.java]: That one's on me. We can discuss offline if you like.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java (hunk 37 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/SymbolProblem.java]: Updating it https://github.com/GoogleCloudPlatform/cloud-opensource-java/pull/602/files#diff-f81d38e11d3d9aa7ef6149c9be9aafbaR64
